### PR TITLE
Migrate from wasmi to wasmer

### LINKED ISSUE #4: Switch to Cranelift
Just opening this up as a soft reminder cause you said you're planning on it! 😄 

https://www.reddit.com/r/PHP/comments/9y7wqw/run_wasm_in_php_natively_php_extension_written_in/e9znbtn/

### PR DESCRIPTION
**Pull request by Hywan**
Fix #4.

This PR replaces `wasmi` by [Wasmer](https://github.com/wasmerio/wasmer) through the `wasmer-runtime-c-api`, and thus we will be able to use Cranelift and soon LLVM for the backend.

### CODE REVIEW

**Comment by Hywan** [comment]
Early results with the n-body benchmark.

Note: I'm testing with PHP 7.3.2.

#### Rust only

```sh
$ hyperfine './nbody-rs 5000000'
Benchmark #1: ./nbody-rs 5000000

  Time (mean ± σ):     287.0 ms ±   8.1 ms    [User: 280.5 ms, System: 2.8 ms]

  Range (min … max):   278.5 ms … 300.5 ms
```

#### Pure PHP

```sh
$ hyperfine 'php nbody.pure.php 5000000'
Benchmark #1: php nbody.pure.php 5000000

  Time (mean ± σ):     19.761 s ±  0.842 s    [User: 19.602 s, System: 0.075 s]

  Range (min … max):   18.908 s … 20.915 s
```

#### `php-ext-wasm` with wasmi

```sh
$ hyperfine 'php -d extension=wasm nbody.wasm.php'
Benchmark #1: php -d extension=wasm nbody.php

 ⠦ Current estimate: 67.622 s     ███████░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░░ ETA 00:18:02
^C
```

I stopped the benchmark, I didn't want to wait 18mn.

#### 🆕 `php-ext-wasm` with wasmer

```sh
$ hyperfine 'php -d extension=wasm nbody.wasm.php'
Benchmark #1: php -d extension=wasm nbody.wasm.php

  Time (mean ± σ):      2.365 s ±  0.266 s    [User: 2.284 s, System: 0.047 s]

  Range (min … max):    2.014 s …  2.864 s
```

#### Conclusion

| Implementation | Time (ms) |
|-|-:|
| Rust (baseline) | 287 |
| Pure PHP | 19'761 |
| `php-ext-wasm` with `wasmi` (current) | 67'622 |
| `php-ext-wasm` with `wasmer` (this PR) | 2365 |

File: extension/wasm.c [hunk 1]
```diff
@@ -284,180 +272,164 @@ PHP_FUNCTION(wasm_new_instance)
...
+
+    if (function_output.tag == WASM_I32) {
+        RETURN_LONG(function_output.value.I32);
+    } else if (function_output.tag == WASM_I64) {
+        RETURN_LONG(function_output.value.I64);
+    } else if (function_output.tag == WASM_F32) {
+        RETURN_DOUBLE(function_output.value.F32);
+    } else if (function_output.tag == WASM_F64) {
+        RETURN_DOUBLE(function_output.value.F64);
     } else {
```
Comment (Ocramius) [inline extension/wasm.c]: Can be omitted, no?

File: src/lib.rs [hunk 2]
```diff
@@ -1,551 +1 @@
...
-            signature,
-            implementation_pointer: *implementation_pointer,
-            zend_fcall_info,
-            zend_fcall_info_cache,
-        };
-
-        wasm_runtime.add_function(function);
-    }
-}
+// Empty file to use Cargo simply.
```
Comment (Ocramius) [inline src/lib.rs]: Why is this simplifying cargo usage?

File: extension/wasm.c [hunk 3]
```diff
@@ -23,31 +23,32 @@
 #include "php.h"
 #include "ext/standard/info.h"
 #include "php_wasm.h"
-#include "php-ext-wasm.h"
+#include "wasmer.h"
 
 /**
- * `wasm_read_binary`.
+ * `wasm_read_bytes`.
```
[the remark this answers is not in the document]
Comment (Hywan) [reply extension/wasm.c]: The PR is a WIP, I'm gonna document more :-).

File: extension/wasm.c [hunk 4]
```diff
@@ -56,187 +57,167 @@ PHP_FUNCTION(wasm_read_binary)
...
-        function_implementation,
-        function_implementation_cache
-    );
-    
-    free(&ffi_signature);
-
-    RETURN_TRUE
-}
+//const Value* invoke_imported_function_from_host(
+//    const /* zend_fcall_info */ void *function_implementation,
```
[the remark this answers is not in the document]
Comment (Hywan) [reply extension/wasm.c]: The PR is a WIP :-).

File: extension/wasm.c [hunk 5]
```diff
@@ -284,180 +272,164 @@ PHP_FUNCTION(wasm_new_instance)
...
-
-    array_init_size(return_value, wasm_function_signature->inputs_length + 1 /* output */);
-
-    for (uintptr_t nth = 0; nth < wasm_function_signature->inputs_length; ++nth) {
-        add_next_index_long(return_value, wasm_function_signature->inputs_buffer[nth]);
-    }
-
-    add_next_index_long(return_value, *(wasm_function_signature->output));
-}
+///**
```
[the remark this answers is not in the document]
Comment (Hywan) [reply extension/wasm.c]: The PR is a WIP :-).

File: extension/wasm.c [hunk 6]
```diff
@@ -284,180 +272,164 @@ PHP_FUNCTION(wasm_new_instance)
...
+    zval *wasm_instance_resource;
+    char *function_name;
+    size_t function_name_length;
+    HashTable *inputs;
 
-    if (zend_parse_parameters(ZEND_NUM_ARGS() TSRMLS_CC, "rd", &wasm_arguments_builder_resource, &number) == FAILURE) {
+    if (
+        zend_parse_parameters(
+            ZEND_NUM_ARGS() TSRMLS_CC,
+            "rsh",
```
[the remark this answers is not in the document]
Comment (Hywan) [reply extension/wasm.c]: `r` for resource,
`s` for string,
`h` for hashtable.

See https://github.com/php/php-src/blob/master/README.PARAMETER_PARSING_API. This is the classical Zend API.

File: extension/wasm.c (hunk 1 above)
Comment (Hywan) [reply extension/wasm.c]: We never know with C!

File: src/lib.rs (hunk 2 above)
Comment (Hywan) [reply src/lib.rs]: Will document more.

**Comment by Hywan** [comment]
Blocked by https://github.com/wasmerio/wasmer/pull/261

File: build.rs [hunk 7]
```diff
@@ -1,11 +1,54 @@
...
+        header_destination.push("wasmer.h");
+        header_destination.set_extension("h");
 
-    cbindgen::generate(crate_dir)
-        .expect("Unable to generate C bindings.")
-        .write_to_file("headers/php-ext-wasm.h");
+        copy(entry.path(), header_destination)
+            .expect("Cannot copy the `wasmer.h` C header file from `wasmer-runtime-c-api`.");
+    } else {
+        panic!("The `wasmer.h` file is not found.");
```
[the remark this answers is not in the document]
Comment (Hywan) [reply build.rs]: The `panic!` is in an `else` because the `if let Some(entry)` is super convenient when written this way :-).

**Pull request merged**